patches/minecraft/net/minecraft/item/BowItem.java.patch

```diff
@@ -1,17 +1,23 @@
 --- a/net/minecraft/item/BowItem.java
 +++ b/net/minecraft/item/BowItem.java
-@@ -24,18 +24,23 @@
+@@ -23,19 +23,25 @@
+       if (p_77615_3_ instanceof PlayerEntity) {
           PlayerEntity playerentity = (PlayerEntity)p_77615_3_;
           boolean flag = playerentity.field_71075_bZ.field_75098_d || EnchantmentHelper.func_77506_a(Enchantments.field_185312_x, p_77615_1_) > 0;
-          ItemStack itemstack = playerentity.func_213356_f(p_77615_1_);
+-         ItemStack itemstack = playerentity.func_213356_f(p_77615_1_);
+-         if (!itemstack.func_190926_b() || flag) {
+-            if (itemstack.func_190926_b()) {
+-               itemstack = new ItemStack(Items.field_151032_g);
++         org.apache.commons.lang3.tuple.Pair<ItemStack, java.util.function.Consumer<ItemStack>> pair = net.minecraftforge.common.ForgeHooks.findAmmo(p_77615_1_, playerentity);
++         ItemStack itemstack = pair.getKey();
 +
 +         int i = this.func_77626_a(p_77615_1_) - p_77615_4_;
 +         i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(p_77615_1_, p_77615_2_, playerentity, i, !itemstack.func_190926_b() || flag);
 +         if (i < 0) return;
 +
-          if (!itemstack.func_190926_b() || flag) {
-             if (itemstack.func_190926_b()) {
-                itemstack = new ItemStack(Items.field_151032_g);
++         if (!pair.getKey().func_190926_b() || flag) {
++            if (pair.getKey().func_190926_b()) {
++               pair = org.apache.commons.lang3.tuple.Pair.of(new ItemStack(Items.field_151032_g), pair.getValue());
```

Comment:
This whole thing would be a lot cleaner if you didn't change those last 3 lines unnecessarily. The vanilla logic is fine here because:
1. `itemstack` is equal to `pair.getKey()` here.
2. `pair.getKey()` is never accessed beyond this point, so this is actually _wrong_ (because `itemstack` is used below).

Reply:
This should be resolved now?